OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/optimobile/OptimovePreferenceCenter.java

```diff
@@ -0,0 +1,265 @@
+                } else {
+                   Preferences preferences = mapResponseToPreferences(response);
+
+                   if (preferences == null) {
+                       // log?
+                       return;
+                   }
+
+                    this.fireCallback(preferences);
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+            } catch (Optimobile.PartialInitialisationException e) {
+                // noop
+            }
+        }
+
+        private void fireCallback(Preferences preferences) {
+            Optimobile.handler.post(() -> GetPreferencesRunnable.this.callback.run(preferences));
+        }
+    }
+
+    static class SetPreferencesRunnable implements Runnable {
+        private final PreferencesSetHandler callback;
+        private final List<PreferenceUpdate> updates;
+
+        SetPreferencesRunnable(PreferencesSetHandler callback, List<PreferenceUpdate> updates) {
+            this.callback = callback;
+            this.updates = updates;
+        }
+
+        @Override
+        public void run() {
+            String mappedRegion = shared.mapRegion(config.getRegion());
+            String url = "https://preference-center-" + mappedRegion + ".optimove.net/api/v1/preferences?customerId=" + customerId + "&brandGroupId=" + config.getBrandGroupId();
+
+            OptimobileHttpClient httpClient = OptimobileHttpClient.getInstance();
+
+            try {
+                //TODO: Need to map Channels to int for updates
+                JSONArray data = new JSONArray(updates.toArray());
+                Response response = httpClient.putSync(url, data);
+
+                if (!response.isSuccessful()) {
+                    this.fireCallback(false);
+                    return;
+                } else {
+                    this.fireCallback(true);
+                    return;
+                }
+            } catch (JSONException | IOException e) {
+                e.printStackTrace();
+            } catch (Optimobile.PartialInitialisationException e) {
+                // noop
+            }
+
+            this.fireCallback(false);
+        }
+
+        private void fireCallback(Boolean result) {
+            Optimobile.handler.post(() -> SetPreferencesRunnable.this.callback.run(result));
+        }
+    }
+
+    private String mapRegion(String region) {
+        switch (region) {
+            case "uk-1":
+                return "dev-pb";
+            case "eu-central-2":
+                return "eu";
+            case "us-east-1":
+                return "us";
+            default:
+                throw new IllegalArgumentException("Region is not supported: " + region);
+        }
+    }
+
+    static void initialize(OptimoveConfig currentConfig, String newCustomerId) {
+        shared = new OptimovePreferenceCenter();
+        config = currentConfig;
+        customerId = newCustomerId;
+    }
+
+    void setCustomerId(@Nullable String newCustomerId) {
```

Comment:
1. think here since all sdk is one, dont need to store a copy of customer id -- more copies === more worries about updating everywhere when customer id changes
2. We can get customer id directly from storage, but we need to make sure it's a customer, not visitor id (if it's a visitor we dont attempt pref requets)
3. and in relation to this , maybe it makes sense to move preference center completely out of optimobile namespace? it would be like in web a separate "module" . PC would lose access to all package-protected classes in optimobile. Currently it uses
- network client, but there is one in com.optimove.android.main.tools.networking;  network client in optimobile can throw PartialInitialisationException when apiKey/secretKey not set. But PC doesnt need these set -- just 1 example how PC is a bit of a standalone thing
- will need to get customer id. getCurrentUserIdentifier exists in optimobile, but that gives you customer or visitor id. VS the one from UserInfo in main is more of what we need
- executor service, but we can have a copy

will check more tomorrow

Reply:
It affects the place PC takes in this world. Now there is Optimove and Optimobile, they are values in FeatureSet enum and are linked to optimove/optimobile credentials. PC doesnt require any credentials and could be 3rd always configured feature

```
    public boolean isPreferenceCenterConfigured() {
        return true;
    }
```

User id can get with `Optimove.getInstance().userInfo.getUserId()` and check it's not equal to visitor id and is not null

@scotthowie-optimove @cgwyllie makes any sense / did i miss something?

Reply:
Concur, PC is not specifically an Optimobile concern, it may manage prefs for other channels / be enabled without an Optimobile channel etc.

Agreed generally preferable not to duplicate/cache state around if possible to avoid. (In Optimobile, we could also explore refactoring this in the future to read directly from appropriate core storage, current approach was the minimum viable integration.)

There may be room to extract a more generic networking client if there isn't already one in core to reuse (sounds like there is?). Probably again a future item to consider refactoring this to common networking code throughout to reduce duplication of effort / dependency footprint (if possible).

---

Re duplication of Executor, we should be mindful of creating too many threads/threadpools in the SDK, perhaps something to refactor or think about in future (not currently sure this is a problem, but just to be aware, we should think a little when adding more executors--it may not only be our SDK creating them and so on). https://developer.android.com/topic/performance/threads#how-many-threads-should-you-create